flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/StronglyConnectedComponentsComputeUtils.java

```diff
@@ -0,0 +1,121 @@
+ */
+public final class StronglyConnectedComponentsComputeUtils {
+
+	static Set<Set<Integer>> computeStronglyConnectedComponents(final int numVertex, final List<List<Integer>> outEdges) {
+		final Set<Set<Integer>> stronglyConnectedComponents = new HashSet<>();
+
+		final int[] vertexIndices = new int[numVertex];
+		Arrays.fill(vertexIndices, -1);
+
+		final int[] vertexLowLinks = new int[numVertex];
+		final Deque<Integer> stack = new ArrayDeque<>(numVertex);
+		final boolean[] onStack = new boolean[numVertex];
+		final AtomicInteger indexCounter = new AtomicInteger(0);
+
+		for (int vertex = 0; vertex < numVertex; vertex++) {
+			if (vertexIndices[vertex] == -1) {
+				dfs(vertex, outEdges, vertexIndices, vertexLowLinks, stack, onStack, indexCounter, stronglyConnectedComponents);
+			}
+		}
+
+		return stronglyConnectedComponents;
+	}
+
+	private static void dfs(
+			final int rootVertex,
+			final List<List<Integer>> outEdges,
+			final int[] vertexIndices,
+			final int[] vertexLowLinks,
+			final Deque<Integer> stack,
+			final boolean[] onStack,
+			final AtomicInteger indexCounter,
+			final Set<Set<Integer>> stronglyConnectedComponents) {
+
+		final Deque<Tuple2<Integer, Integer>> loopStack = new ArrayDeque<>();
+		loopStack.add(new Tuple2<>(rootVertex, 0));
+
+		while (!loopStack.isEmpty()) {
+			Tuple2<Integer, Integer> tuple = loopStack.pollLast();
+			final int currentVertex = tuple.f0;
+			final int vertexLoopCount = tuple.f1;
+
+			if (vertexLoopCount == 0) {
+				vertexIndices[currentVertex] = indexCounter.get();
+				vertexLowLinks[currentVertex] = indexCounter.getAndIncrement();
+				stack.add(currentVertex);
+				onStack[currentVertex] = true;
+			} else if (vertexLoopCount > 0) {
+				final int successorVertex = outEdges.get(currentVertex).get(vertexLoopCount - 1);
+				vertexLowLinks[currentVertex] = Math.min(vertexLowLinks[currentVertex], vertexLowLinks[successorVertex]);
+			}
+
+			boolean visitSuccessorVertex = false;
+			for (int i = vertexLoopCount; i < outEdges.get(currentVertex).size(); i++) {
+				final int successorVertex = outEdges.get(currentVertex).get(i);
+				if (vertexIndices[successorVertex] == -1) {
+					loopStack.add(new Tuple2<>(currentVertex, i + 1));
+					loopStack.add(new Tuple2<>(successorVertex, 0));
+					visitSuccessorVertex = true;
+					break;
+				} else if (onStack[successorVertex]) {
+					vertexLowLinks[currentVertex] = Math.min(vertexLowLinks[currentVertex], vertexIndices[successorVertex]);
+				}
+			}
+
+			if (visitSuccessorVertex) {
+				continue;
+			}
+
+			if (vertexLowLinks[currentVertex] == vertexIndices[currentVertex]) {
+				final Set<Integer> scc = new HashSet<>();
+				while (onStack[currentVertex]) {
+					final int v = stack.pollLast();
+					onStack[v] = false;
+					scc.add(v);
+
+				}
+				stronglyConnectedComponents.add(scc);
+			}
+		}
+	}
+
+	private StronglyConnectedComponentsComputeUtils() {
```

Comment:
From what I saw before and discussed as code style, we usually put instance methods before static methods.